packages/core/contracts/proxy-scripts/lsp-broker/LspUniswapV2Broker.sol

```diff
@@ -19,23 +21,55 @@ contract LspUniswapV2Broker {
 
     /**
      * @notice Mint long and short tokens and deposit them all into a UniV2 Pool.
+
+     * @param callingAsEOA If True, caller has balance of collateral and expects to receive back all long/short tokens.
+     * @param longShortPair LSP contract address to mint position on.
+     * @param router Contract to call to exchange long and short tokens.
+     * @param amountCollateral Amount of collateral to deposit and mint long and short tokens against.
      */
-    function atomicMintAddLiquidity() public {
-        /* TODO */
+    function atomicMintAddLiquidity(
+        bool callingAsEOA,
+        LongShortPair longShortPair,
+        IUniswapV2Router01 router,
+        uint256 amountCollateral
+    ) public {
+        require(address(longShortPair) != address(0), "Invalid long short pair");
+        require(address(router) != address(0), "Invalid router");
+        require(amountCollateral != 0, "Collateral to mint with");
+
+        IERC20 collateralToken = IERC20(longShortPair.collateralToken());
+        IERC20 longToken = IERC20(longShortPair.longToken());
+        IERC20 shortToken = IERC20(longShortPair.shortToken());
+
+        // 0) Pull collateral from caller if necessary and approve LSP to spend it.
+        if (callingAsEOA)
+            TransferHelper.safeTransferFrom(address(collateralToken), msg.sender, address(this), amountCollateral);
+
+        TransferHelper.safeApprove(address(collateralToken), address(longShortPair), amountCollateral);
+
+        // 1) Deposit collateral into LSP and mint long and short tokens.
+        longShortPair.create(amountCollateral);
+
+        // 2) Send long and short tokens to the pair address and call the low-level mint function. This amounts to
+        // single asset deposit where the one side will be sold for the other to match the pool ratio.
+        IUniswapV2Pair pair = pairFor(router.factory(), longToken, shortToken);
+        TransferHelper.safeTransfer(address(longToken), address(pair), longToken.balanceOf(address(this)));
+        TransferHelper.safeTransfer(address(shortToken), address(pair), shortToken.balanceOf(address(this)));
+        pair.mint(callingAsEOA ? msg.sender : address(this));
```

Comment:
Does `mint` guarantee that it uses 100% of the two tokens to create LP tokens? I.e. there is no dust?

Reply:
I'm pretty sure it does, ye